### DISCUSSION TITLE
Better data structure for a Grid

### DISCUSSION THREAD

**Discussion by sbenthall** [scikit-agent/scikit-agent#48] [Ideas]
We have the need for a data structure that's something like a matrix or tensor with labeled columns.
A seed of this functionality is in `grid.py` but this can be expanded.

Generating the grid from a configuration of the range for each dimension is currently in place, but this doesn't create an object with labels.

(Note that this is much like an xarray and one day might be replaced by it, but we need this to wrap a PyTorch tensor. I'm inclined to build what we need first and then swap in xarray later with a light wrapper if that works, because a 'grid' is not quite the same as an xarray (it's the 'coordinates' of the xarray without the values assigned to those coordinates....  I suppose it's also like a pandas dataframe?)

Other features:

- We'll need to be able to pass in additional data points (such as a 'natural borrowing constraint' for a discretized transitory income shock, in the `SolvingMicroDSOPs` style of problem) -- ideally these can be mathematical expressions and evaluated in scope
- methods for changing the grid into other representations of the same data, when that's useful (i.e., a dict with keys as labels and one-dimensional tensors as values)

@alanlujan91 I wonder about your ideas about this since you've had experience with xarray-based solution frameworks before.

**Comment by sbenthall** [comment]
I'm thinking:

- Grid class
- initialized with a config; grid made to match
- i guess keep the data as an attribute for now (it would be nice if it could be a subclass of arrays/tensors, but it's hard to be both)
- make switching to torch mode part of the methods/initialization
- the logic for decomposing it into a dictionary is currently in `get_estimated_discounted_lifetime_reward_loss` and can be moved

**Comment by alanlujan91** [comment]
something like what i had in labeled models would be ideal 
https://github.com/econ-ark/HARK/blob/master/HARK/ConsumptionSaving/ConsLabeledModel.py

```
u = UtilityFuncCRRA(CRRA)

    mNrm = xr.DataArray(
        np.append(0.0, aXtraGrid),
        name="mNrm",
        dims=("mNrm"),
        attrs={"long_name": "cash_on_hand"},
    )
    state = xr.Dataset({"mNrm": mNrm})  # only one state var in this model

    # optimal decision is to consume everything in the last period
    cNrm = xr.DataArray(
        mNrm,
        name="cNrm",
        dims=state.dims,
        coords=state.coords,
        attrs={"long_name": "consumption"},
    )

    v = u(cNrm)
    v.name = "v"
    v.attrs = {"long_name": "value function"}

    v_der = u.der(cNrm)
    v_der.name = "v_der"
    v_der.attrs = {"long_name": "marginal value function"}

    v_inv = cNrm.copy()
    v_inv.name = "v_inv"
    v_inv.attrs = {"long_name": "inverse value function"}

    v_der_inv = cNrm.copy()
    v_der_inv.name = "v_der_inv"
    v_der_inv.attrs = {"long_name": "inverse marginal value function"}

    dataset = xr.Dataset(
        {
            "cNrm": cNrm,
            "v": v,
            "v_der": v_der,
            "v_inv": v_inv,
            "v_der_inv": v_der_inv,
        }
    )
```

**Comment by alanlujan91** [comment]
Also see https://github.com/econ-ark/HARK/blob/master/examples/LabeledModels/LabeledModels.ipynb

**Comment by alanlujan91** [comment]
Something that hasn't come up in our use, but I've been thinking about a lot, is "curvilinear" grids or unstructured grids, which I don't think are supported by xarray 

If that issue ever comes up, we could look into https://github.com/UXARRAY/uxarray

**Comment by alanlujan91** [comment]
Some discussion about xarray+torch here https://github.com/pydata/xarray/issues/3232

something we could do "now" is numpy structured arrays to/from tensor dicts

https://docs.pytorch.org/tensordict/main/reference/generated/tensordict.from_struct_array.html
https://docs.pytorch.org/tensordict/stable/index.html
https://numpy.org/doc/stable/user/basics.rec.html

**Comment by sbenthall** [comment]
A version of this was implemented in #45 -- the Grid class now has the matrix/tensor content but also acts in a 'labeled' way.

We can discuss other potential changes here and then make more specific issues.

**Comment by MridulS** [comment]
Was playing around with the pytorch bits, and why is scikit-agent even mixing numpy into the code while building the neural nets? pytorch tensors work on both cpu and gpu, so the whole numpy codepath should be avoidable?

And as Alan mentioned `TensorDict` looks like a good data structure for the Grid?

**Reply by sbenthall** [reply]
You're making a great point but....

I don't think we've committed to scikit-agent being _only_ about neural net based solutions.
Grid objects are useful for both neural network and traditional dynamic programming approaches, and so I was trying to find a way to be compatible with both torch and numpy.

That might be too confusing in the long run, or maybe it makes sense to have the backing data structure be a `TensorDict` in all cases and do the type conversions when necessary.

I'm open to all these options and designs.

**Reply by MridulS** [reply]
> That might be too confusing in the long run, or maybe it makes sense to have the backing data structure be a TensorDict in all cases and do the type conversions when necessary.

I would argue, maybe it's not even required to convert to numpy at all?

pytorch supports [the array api standard](https://data-apis.org/array-api/latest/) and it should cover most of the functionality. Even for scipy specific things, pytorch can be used via https://scipy.github.io/devdocs/dev/api-dev/array_api.html#using-array-api-standard-support

**Reply by sbenthall** [reply]
Awesome! When will the array API standard support 'labeled' columns?

**Reply by MridulS** [reply]
> array API standard support 'labeled' columns

I don't think that's happening any time soon. There have been some efforts https://discuss.scientific-python.org/t/multiple-duck-array-layers-composability-implementations-and-usability-concerns/552 but this will be outside of the array API standard, as this is out of scope.
